### PR TITLE
[aspnet] Add url.path to sampler tags

### PR DESCRIPTION
**Pull request by Wraith2**
Added the "url.path" tag to the tags for newly created activities in StartAspNetActivity.

To ignore a path like a healthcheck it is best to use a sampler to ensure that the activity and all child activities are dropped in a sampler. To do this the sampler needs to be able to provide enough information for the sampler author to detect the path. This is only currently possible by using HttpContext.Current or other method of injecting the request into the sampler. This is unsatisfying because the caller of the sampler has information about the current request it just doesn't have a way to pass that information to the sampler. 

This PR changes the way StartAspNetActivity works by adding the request.Unvalidated.Path of the request to the tags that are used to construct the activity. The tags are one of the few things that are passed to samplers when an activity is started. With this change it is now possible to write a sampler which uses only provided context information in the sampler to decide whether a particular request root activity should be dropped.

There are no public api changes.

### CODE REVIEW

**Comment by martinjt** [comment]
I think as we're adding this, it would be better to add as much as we can, not just the path, anything we have that's available at that time.

What about the code that adds these properties later, should they be removed?

I'm also curious about whether constants for the keys would be better.

**Comment by Wraith2** [comment]
> I'm also curious about whether constants for the keys would be better.

Using one of the existing constants would require that I add a project reference to the OpenTelemetry.SemanticConventions assembly. Adding a local constant seemed pointless when this is the only use of it in this library, if it were multiuse in this project I'd have made it a constant. If you prefer a constant or want me to add the assembly dependency let me know which you'd like.

> I think as we're adding this, it would be better to add as much as we can, not just the path, anything we have that's available at that time.

Very little is available at this time because it's very early in the request lifecycle. No validation or parsing has been done yet which is why I'm using the Unvalidated property of the request. All the attributes which are added later are added by the AspNet integration not by the telemetry module and they require various amounts of validation and parsing. The only other possibly useful information I can think of would the the route that is used but it isn't and can't be made available at this time in the lifecycle. Things like http version, http method really don't seem like useful attributes for making a sampling decision.

File: src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs:68]: instrumentations are generally not doing this due to concern over heap allocation. This must, however, be solved, and need to work with the owners (ActivitySource/Activity i.e runtime team) to see if they can support `TagList` or `StartActivity` overloads for taking 1,2,3 tags, so this allocation can be avoided.

Comment (Wraith2) [reply src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs:68]: I couldn't find a good way to deal with the allocation The api requires IEnumerable so whatever happens it either has be a class or it'll be boxed. I tried TagList but decided the boxing of such a large struct was going to be more expensive than  what the compiler is likely to do here, which is probably a new single length array. I considered creating a class which can contain the single element and which is enumerable but it felt like overkill for a first contribution to the project.

Comment (cijothomas) [reply src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs:68]: it is not possible to avoid this, without changes in the API offered by runtime itself.

The only instrumentation in the repo providing tags at startup is the sqlclient one, but only does so for fully static tag: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/src/OpenTelemetry.Instrumentation.SqlClient/Implementation/SqlActivitySourceHelper.cs#L25

Comment (cijothomas) [reply src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs:68]: > instrumentations are generally not doing this due to concern over heap allocation. This must, however, be solved,

@Wraith2  Just to make it clear - I didn't mean that this should be solved in this PR or you should work with runtime team to solve this. Just stated that this is a problem that must be solved. (I can create a tracking issue in runtime and otel repos, once i find the old discussion/probably issue itself asking the same)

OTel convention requires a lot of tags to be provided at startup time itself, and it is not possible to do it without allocation in Activity, so this is a problem that needs to be solved even outside of this issue/pr.

Comment (cijothomas) [reply src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs:68]: Couldn't find a tracking issue, but https://github.com/dotnet/runtime/issues/42784#issuecomment-733705736 has explained the problem. If you are up for it, please create a new issue in the runtime repo. Else, I can create one.

(Also, it is very unlikely that runtime will be able to make any changes for .NET 9 timeframe, so we are looking at .NET 10 timeframe only at this point)

Comment (Wraith2) [reply src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs:68]: New proposal for a CreateActivity overload that takes a single kvp, https://github.com/dotnet/runtime/issues/103245

**Comment by cijothomas** [comment]
> Just for clarity, what is this waiting on? and if it's me can you tell me what is needed?

I'll let others review, but my concern is the allocation in hot path. It _might be okay_ because of the fact that `Activity` itself will be created (heap allocated) irrespective of sampling decision (as it is root), so adding few more bytes (for the tag) may not be critical. Also once runtime supports passing tags without allocation (in .NET 10 timeframe), this won't be an issue.

File: src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs:68]: is this tag also populated by the aspnet instrumentation as well?

Comment (cijothomas) [reply src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs:68]: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/src/OpenTelemetry.Instrumentation.AspNet/Implementation/HttpInListener.cs#L87C67-L87C80 can be removed, if we proceed with this approach.

Comment (Wraith2) [reply src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs:68]: It can only be removed if it is required that TelemetryHttpModule be used with AspNet. It's possible to use AspNet without the telemetry module and in that case users would be missing the `url.path` attribute.

Comment (cijothomas) [reply src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs:68]: > if it is required that TelemetryHttpModule be used with AspNet.

It is [required!](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/tree/main/src/OpenTelemetry.Instrumentation.AspNet#step-2-modify-webconfig)

Comment (Wraith2) [reply src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs:68]: Ok. removed from AspNet because it's now set in the telemetry module. That was the only unvalidated property being accessed.

**Comment by Wraith2** [comment]
> > 'm also curious about whether constants for the keys would be better.
> 
> Using one of the existing constants would require that I add a project reference to the OpenTelemetry.SemanticConventions

Looking at AspNet is includes SemanticConventions by linking to the file. Is there a specific reason for this? I expected there to be a nuget package containing public versions of the convention constants but there doesn't seem to be one.
I can add a link to the shared file and use the convention constant definitions but it seems like bloat to do so.

**Comment by martinjt** [comment]
We're working on a method for producing the conventions long-term and what that will look like, but's not ready yet. It's an ongoing task for all the languages.

**Comment by martinjt** [comment]
In terms of the wider idea of the allocations, my suggestion to sidestep this if it's a concern is to make it opt-in by adding a parameter that will do it instead of doing it by default.

**Comment by Wraith2** [comment]
That's a pretty aggressive stale marking behaviour. Is there a specific thing blocking this PR?

**Comment by CodeBlanch** [comment]
> This is unavoidable due to conflicting rules. One rule requires a space and another requires there to be no space so both cannot be satified. `enumerable = [new KeyValuePair<string, object?>("url.path", path)];` gives `SA1010 An opening square bracket within a C# statement is not spaced correctly.` applying the fixer gives: `enumerable =[new KeyValuePair<string, object?>("url.path", path)];` gives `SA1003 The spacing around an operator symbol is incorrect, within a C# code file`

@Wraith2 

Bump this...

https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/e3a90d7ac5b0cb50ebe9838c87a6e6fe06d4dd40/build/Common.props#L46

...to...

```diff
-    <StyleCopAnalyzersPkgVer>[1.2.0-beta.507,2.0)</StyleCopAnalyzersPkgVer>
+    <StyleCopAnalyzersPkgVer>[1.2.0-beta.556,2.0)</StyleCopAnalyzersPkgVer>
```

The new StyleCop version seems to do the right thing and stops warning.

File: src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs
Comment (CodeBlanch) [reply src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs:68]: Could we use a `ThreadStatic` for the storage?

```csharp
    [ThreadStatic]
    private static KeyValuePair<string, object?>[]? samplerTagStorage;

        IEnumerable<KeyValuePair<string, object?>>? tags;
        if (context.Request?.Unvalidated?.Path is string path)
        {
            var samplerTags = samplerTagStorage ??= new KeyValuePair<string, object?>[1];
            samplerTags[0] = new KeyValuePair<string, object?>("url.path", path);
            tags = samplerTags;
        }
        else
        {
            tags = null;
        }

        Activity? activity = AspNetSource.StartActivity(TelemetryHttpModule.AspNetActivityName, ActivityKind.Server, propagationContext.ActivityContext, tags);
```

Pretty sure sampling happens on the same thread so it is mostly safe. Some user could make a custom sampler that attempts to hold onto the tags but why would they?

**Comment by CodeBlanch** [comment]
@martinjt 

> In terms of the wider idea of the allocations, my suggestion to sidestep this if it's a concern is to make it opt-in by adding a parameter that will do it instead of doing it by default.

Seems like making it an opt-in thing via options is reasonable. What would that option look like though? The [spec](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/http/http-spans.md#http-server) defines a bunch of tags for the sampler:

> [client.address](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/attributes-registry/client.md)
[http.request.header.<key>](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/attributes-registry/http.md)
[http.request.method](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/attributes-registry/http.md)
[server.address](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/attributes-registry/server.md)
[server.port](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/attributes-registry/server.md)
[url.path](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/attributes-registry/url.md)
[url.query](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/attributes-registry/url.md)
[url.scheme](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/attributes-registry/url.md)
[user_agent.original](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/attributes-registry/user-agent.md)

We probably don't want an option for everything. Could be a flags enum or `HashSet` or something?

File: src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs
Comment (Wraith2) [reply src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs]: The tags will get added to the TagList in the activity and since KeyValuePair is a struct they're copied and the reference to the incoming array is not retained. So it'll be safe as long as no await is added along the current path.

I've added this change.

File: src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs:81]: is there a need to do this reset?

**Review by cijothomas — COMMENTED** [review]
LGTM.
Please add changelog entry, given this is end user facing improvement.
Also, it'd be good to see if this can be covered via unit tests, esp. the TLocal part.

File: src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs
Comment (Wraith2) [reply src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs:81]: Need? You could argue that it isn't essential to the functionality. 
I think it is a good idea to cleanup the value so that we don't have a reference to some random string living after it's been used. The code required to check and clear the field is a handful of assembly instructions.

**Comment by Wraith2** [comment]
I've added a changelog entry.

I looked at how to add tests and couldn't find a good way to do it. The Opentelemetry.Trace namespace isn't available in the HttpInListener test project and it isn't clear how to add the reference to the correct version from the other projects around it.
The presence of `url.path` is already tested in the existing asp.net tests.

**Comment by Wraith2** [comment]
Disappointing.

**Comment by cijothomas** [comment]
> Disappointing.

Don't be! The automation just closes PRs if not active for 7 days! I'll reopen now and request reviews again. (Some of the maintainers were on break for last few days, but mostly back now I think.)

**Review by CodeBlanch — APPROVED** [review]
LGTM

I tweaked the CHANGELOG and added a test.

@vishweshbankwar @Kielek Would one of you please review as well? I don't want to self-approve anything 😄

File: src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/CHANGELOG.md
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/CHANGELOG.md:8]:
```suggestion
  available to samplers and may be used to influence the sampling by authoring custom samplers
```

Comment (cijothomas) [reply src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/CHANGELOG.md:8]: Suggetsion: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/tree/main/src/OpenTelemetry.Instrumentation.AspNet#trace-filter can mention that it is recommended to do sampling wherever feasible, and state that url.path is part of attributes available for samplers.

**Review by cijothomas — APPROVED** [review]
LGTM. Left couple of nit suggestions, not blockers.

File: src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs
Comment (vishweshbankwar) [inline src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/ActivityHelper.cs:69]: can this be simplified?

`cachedTagsStorage ??= new KeyValuePair<string, object?>[1];`

**Review by vishweshbankwar — APPROVED** [review]
LGTM

File: src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/CHANGELOG.md
Comment (CodeBlanch) [reply src/OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule/CHANGELOG.md]: @cijothomas I pushed some changes. LMK what you think.

File: src/OpenTelemetry.Instrumentation.AspNet/README.md
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.AspNet/README.md:147]:
```suggestion
perform filtering based on "url.path" Tag, as it will prevent child spans from being created and bypass
```

Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.AspNet/README.md:147]:
```suggestion
perform filtering as it can ensure consistent sampling when ParentBased samplers are used, by consistently recording/not-recording this Activity and the children..
```

Not sure of how best to phrase this. Will come back later. This is reasonably good to merge now.

**Comment by roycald245** [comment]
> > I'm also curious about whether constants for the keys would be better.
> 
> Using one of the existing constants would require that I add a project reference to the OpenTelemetry.SemanticConventions assembly. Adding a local constant seemed pointless when this is the only use of it in this library, if it were multiuse in this project I'd have made it a constant. If you prefer a constant or want me to add the assembly dependency let me know which you'd like.
> 
> > I think as we're adding this, it would be better to add as much as we can, not just the path, anything we have that's available at that time.
> 
> Very little is available at this time because it's very early in the request lifecycle. No validation or parsing has been done yet which is why I'm using the Unvalidated property of the request. All the attributes which are added later are added by the AspNet integration not by the telemetry module and they require various amounts of validation and parsing. The only other possibly useful information I can think of would the the route that is used but it isn't and can't be made available at this time in the lifecycle. Things like http version, http method really don't seem like useful attributes for making a sampling decision.

Actually it would be nice to have the http method :)
I am in need of the ability to have the http properties in the sampler. but the difference is that i require the http method in addition to the url. Would be nice if this can be added to this PR and not to make another just for another property :)

**Comment by cijothomas** [comment]
@roycald245 It is reasonable to add more tags, esp. since semantic convention recommends that. It requires some [opt-in design model](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1871#issuecomment-2189843381), that is not yet in. We can track it separately (please open an issue), to not delay this PR.

**Comment by roycald245** [comment]
@cijothomas opened #57287. Will open PR as soon as this merges

**Pull request merged**
